### PR TITLE
Printer integration

### PR DESCRIPTION
**Pull request by lu16j**
DO NOT MERGE until corresponding HQ feature is implemented (ticket submitted by Clayton).

TODOs when HQ feature is implemented:
- Remove TEMPLATE_FILE_HACK_EXT and all dependencies on it from TemplatePrinterActivity.
- Update TEMPLATE_FILE_PATH if the template document file will always be in the same location. If not, write a function to determine where the template file is.

Other nice-to-dos:
- Make this a question type rather than a callout.

------

Adds an internal callout activity (used in the same way as an app callout) which loads a .docx/.odt file template, populates it with CommCare data, and sends it off to an external document reader for printing/saving to PDF.

Template files formatting:
- Text occurrences of "{{ attr_key }}" will be replaced with the corresponding attr_value included in the activity's extras.
- The formatting of attr_key will be preserved (i.e. if it was bold and italicized, the data it will be replaced by will also be bold and italicized).

External app dependencies:
- Required: one of
 - Quickoffice or equivalent for .docx templates (recommended; better editors, better viewers)
 - OpenDocument Reader or equivalent for .odt templates
- Optional:
 - File Explorer or equivalent file manager capable of file selection. If template file is not bundled with application, file manager will be launched to manually select a template file.

### CODE REVIEW

**Comment by amstone326** [comment]
@lu16j You originally commented not to merge this until the HQ side was done. Can we amend that now though? Not sure if I should mark this ready for review or not, but in theory it is ready.

**Comment by lu16j** [comment]
@amstone326 When I implemented this, we were waiting on HQ to add functionality to supply a template .docx/.odt file at a specified path. If that is no longer the case, then it should be good to merge, though you may want to consult @wpride who was my manager during my internship.

**Comment by amstone326** [comment]
@lu16j I'm so sorry, I completely blanked on who that tag was. Meant that question for someone else. Sorry that didn't make much sense!

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        // Create a print job with name and adapter instance
+        printJob = printManager.print(jobName, printAdapter,
+                new PrintAttributes.Builder().build());
+        jobStarted = true;
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        if (jobStarted) {
```
Comment (amstone326) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: I'll just say this preemptively- I know it's kind of hacky to have a check in onResume like this, but it was the only way I could come up with for knowing when to finish the activity. Unlike when we were printing a PDF and creating a custom PrintDocumentAdapter for it, a WebView (very conveniently) has a built-in method for creating its own adapter (SEE v.createPrintDocumentAdapter() above). However, this removes the opportunity to place any sort of listener for the result of the job within the onFinish() method of the adapter. 

To give some context of why this works and why I'm confident it doesn't break anything- when the print job is started in createWebPrintJob(), an external Android print dialog pops up where the user can choose what printer to print to, save to PDF, etc. At that point, a return to the print activity can be due to only one of 2 things: 1) the print dialog closes naturally, signaling the print job has ended, or 2) the user presses back. In either case, the desired behavior at that point is for the activity to end (after telling the user what happened with the alert dialog). Technically, there is a 3rd way to end up back at the print activity, and that would be to rotate the screen. However, since this causes the activity to start up entirely new anyway (from onCreate() rather than on Resume()), this has no effect.

All that being said, if anyone has a cleaner idea for how to handle this, let me know!

File: app/res/values/strings.xml [hunk 2]
```diff
@@ -197,6 +197,8 @@ Copyright (c) 2012 readyState Software Ltd.</string>
     <string name="noselect_error">Sorry, you have not selected any forms!</string>
     <string name="no_capture">Nothing recorded yet.</string>
     <string name="no_connection">No network connection available</string>
+    <string name="no_data">No data has been provided for your print activity. Please see a
```
Comment (wpride) [inline app/res/values/strings.xml]: why is this on two lines?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    /**
+     * If an error occurs, stores which type the error was, to be passed back to
+     * TemplatePrinterActivity through the PopulateListener's onError method
+     */
+    private int errorType;
+
+    /**
+     * The 2 types of errors that can be encountered in this task
+     */
+    public static final int IO_ERROR = 0;
```
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: why is one public one private?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+     * Attempts to perform population of the template file, and throws the appropriate exception
+     * if encountering an error.
+     */
+    @Override
+    protected Boolean doInBackground(Void... params) {
+        try {
+            populateHtml(inputFile, values);
+            return true;
+        } catch (IOException e) {
+            errorType = IO_ERROR;
```
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: seems odd having this errorType state-ness... maybe this should just return an integer with EG 0 being success, 1 being IO error, etc.?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        String beforeBodySection = fileText.substring(0, startBodyIndex);
+        String bodySection = fileText.substring(startBodyIndex);
+        validateStringOrThrowException(bodySection);
+
+        // Swap out place-holder keywords for case property values within <body></body> section
+        bodySection = replace(bodySection, values);
+
+        // Write the new HTML to the desired  temp file location
+        // throws IO Exception
+        TemplatePrinterUtils.writeStringToFile(beforeBodySection + bodySection, outputPath);
```
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: So we need an affordance for non-template HTML before the <body> but not after?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    /**
+     * Validates the input string for well-formed {{ }} and < > pairs.
+     * If malformed, throws a RuntimeException which will be caught by
+     * doInBackground() and passed to the attached PopulateListener.onError()
+     * 
+     * @param input String to validate
+     */
+    private static void validateStringOrThrowException(String input) {
```
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Is it possible that we want to pass up some more information to the UI level like which pair was malformed or the approximate location?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+        boolean isBetweenMustaches = false;
+        boolean isBetweenChevrons = false;
+        boolean isWellFormed = true;
+
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+
+            if (c == '{') {
```
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: this function is cool

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+     * Reads back from the encrypted file generated by the above and returns a string
+     * representation of the file's contents
+     *
+     * @param readFromPath the path to the file from which to read
+     * @return a string representation of the file's contents
+     */
+    public static String readStringFromFile(String readFromPath) throws IOException {
+        try {
+            Cipher decrypter = Cipher.getInstance("AES");
+            decrypter.init(Cipher.DECRYPT_MODE, KEY);
```
Comment (wpride) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Ugh... this code is all over the place. Likely out of scope for this feature, but would be great to refactor at some point.

https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/android/db/legacy/LegacyInstallUtils.java#L385
https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/android/crypt/CryptUtil.java#L207
https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/android/tasks/FormRecordCleanupTask.java#L235
https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/dalvik/services/CommCareSessionService.java#L133
https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/android/util/bitcache/FileBitCache.java#L80

Comment (wpride) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: actually I think you might be able to straight up use CryptUtil here, but the file/cipher input stream would still be good to refactor

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+     * printed by the Android print framework
+     *
+     * Source: https://developer.android.com/training/printing/html-docs.html
+     */
+    private void doHtmlPrint() {
+        // Create a WebView object specifically for printing
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(new WebViewClient() {
+
+            public boolean shouldOverrideUrlLoading(WebView view, String url) {
```
Comment (wpride) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: what's this for?

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java (hunk 1 above)
Comment (wpride) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: check here for the pattern you're looking for:

http://stackoverflow.com/questions/30742051/android-printmanager-get-callback

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 6 above)
Comment (lu16j) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Yeah, not entirely sure why I used a boolean flag and a break statement. I would suggest at every place where there's currently a "isWellFormed = false; break;", you could throw an exception with details about placement and type of pattern.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 7 above)
Comment (lu16j) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Ha, thanks. Hope it makes sense!

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 3 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Because one is used externally and one isn't.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 5 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Not sure what you're asking, can you clarify?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 4 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Makes sense, changing.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 5 above)
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Just curious about the "beforeBodySection" affordance - what are we accounting for with this?

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 3 above)
Comment (wpride) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: sorry for not  catching that myself, thanks

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java (hunk 8 above)
Comment (amstone326) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Hm yea kinda figured as I was doing it that I must be duplicating existing stuff, but also needed this to do something pretty specific in terms of the cipher stream. Happy to do a refactor that puts common stuff into a util class, but would definitely want that to be its own PR.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 5 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: The <body></body> section of the html doc is the only place where a user can put any content, so it's only necessary to call replace() on that section of the document. Therefore calling either validateString() or replace() on the entire file string is a waste of time (the validate method is pretty slow since it goes character by character), and also unnecessarily introduces the possibility of having a validation error thrown when we don't want it. Even if there was something that we consider "illegal" in the <head> section, we don't need to care about it, so it doesn't make sense to allow that to happen.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+     * @return The populated String
+     */
+    private static String replace(String input, Bundle values) {
+        // Split input into tokens bounded by {{ and }}
+        String[] tokens = TemplatePrinterUtils.splitKeepDelimiter(input, "\\{{2}", "\\}{2}");
+        for (int i=0; i<tokens.length; i++) {
+            String token = tokens[i];
+
+            // Every 2nd token is a attribute enclosed in {{ }}
+            if (i % 2 == 1) {
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Quoting this warning from the FindBugs Android Studio plugin:
The code uses x % 2 == 1 to check to see if a value is odd, but this won't work for negative numbers (e.g., (-5) % 2 == -1). If this code is intending to check for oddness, consider using x & 1 == 1, or x % 2 != 0.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+ */
+public class TemplatePrinterTask extends AsyncTask<Void, Void, Integer> {
+
+    /**
+     * The 4 result codes that can be sent back by this task
+     */
+    public static final int SUCCESS = 0;
+    public static final int IO_ERROR = 1;
+    public static final int VALIDATION_ERROR_MUSTACHE = 2;
+    public static final int VALIDATION_ERROR_CHEVRON = 3;
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Would be sweet if these were an enum. It would also _slightly_ clean up the TemplatePrinterActivity.onFinished switch statement

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * Used to track which type of error (2 or 3 above) was encountered in validateString
+     */
+    private static int validationErrorType;
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Is there a reason this is static? If there are ever 2 async template tasks running at once, this will be problematic.
If you refactor exception raising/handling (I'll point out a good way in a bit) you probably don't even need this field.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+            recentString.append(c);
+
+            if (recentString.length() > 40) {
+                recentString.deleteCharAt(0);
+            }
+
+            if (c == '{') {
+                if (isBetweenMustaches) {
+                    validationErrorType = VALIDATION_ERROR_MUSTACHE;
+                    throw new RuntimeException(recentString.toString());
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: You should create an explicit exception class that you then raise here and catch it in doInBackground etc. It is bad to raise a generic error, because other code might also produce such errors, so when you catch them later you have no idea where the exception came from.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * The template file for this print action
+     */
+    private final File inputFile;
+
+    /**
+     * The path where the populated template should be saved to
+     */
+    private final String outputPath;
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: I'm a fan of using self-documenting names that make comments unnecessary. Do what you prefer, but maybe consider something like 'inputFile' -> 'templateFile' and 'outputPath' -> 'populatedTemplateFilePath'

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * The mapping from keywords to case property values to be used in populating the template
+     */
+    private final Bundle values;
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: values is super generic, please rename to something like templateSubstitutionMapping

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+    public TemplatePrinterTask(File input, String outputPath, Bundle values,
+                               PopulateListener listener) {
+        this.inputFile = input;
+        this.outputPath = outputPath;
+        this.values = values;
+        this.listener = listener;
+    }
+
+    /**
+     * Attempts to perform population of the template file, and throws the appropriate exception
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Exceptions aren't thrown by this method, contrary to what this comment says.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * Populates an html print template based on the given set of key-value pairings
+     * and save the newly-populated template to a temp location
+     *
+     * @param input the html print template
+     * @param values the mapping of keywords to case property values
+     * @throws IOException
+     */
+    private void populateHtml(File input, Bundle values) throws IOException {
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Would be nice to also take in the output file path, enabling this to be a static method. Also rename to something that indicates that an output file is being written, or do that logic in another method.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+        // Read from input file
+        // throws IOException
+        String fileText = TemplatePrinterUtils.docToString(input).toLowerCase();
+
+        // Check if <body></body> section of html string is properly formed
+        // throws RuntimeException
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Comments that describe what the code does are dangerous; what happens when the code changes? In this case it is easy to see what throws the exceptions because the methods used will have 'throws IOException' annotations on their declarations

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+            throw new RuntimeException(e);
+        } catch (NoSuchPaddingException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        } catch (InvalidKeyException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
```
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Since all of these catches have the same body, you can squash them by using new java 7 syntax:
```
catch (Exception1 | Exception2 | ... | ExceptionN e) { ...
```

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+        in.close();
+        return builder.toString();
+    }
+
+    /**
+     * Writes the given string, encrypted, to the file location specified
+     *
+     * @param fileText the content to write
+     * @param outputPath the path of the file to write to
+     * @throws IOException
```
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Don't include these javadoc annotations if they don't have any content to them.

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+        try {
+            Cipher decrypter = Cipher.getInstance("AES");
+            decrypter.init(Cipher.DECRYPT_MODE, KEY);
+            FileInputStream fis = new FileInputStream(new File(readFromPath));
+            CipherInputStream cis = new CipherInputStream(fis, decrypter);
+            BufferedReader reader = new BufferedReader(new InputStreamReader(cis));
+            String docString = "";
+            String line;
+            while ((line = reader.readLine()) != null) {
+                docString += line;
```
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Instead of += on Strings, you should use the more efficient pattern of StringBuilder and append, like in the 'docToString' method.

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+    private static final String FORMAT_REGEX_WITH_DELIMITER = "((?<=%2$s)|(?=%1$s))";
+    private static final SecretKey KEY = CryptUtil.generateSemiRandomKey();
+
+    /**
+     * Gets the file extension from the given file path.
+     *
+     * @param path File path
+     * @return File extension
+     */
+    public static String getExtension(String path) {
```
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Please avoid obvious javadoc comments. If you rename this method to getFileExtension and the param to filepath, the comments aren't needed.

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+        return last(path.split("\\."));
```
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Needs error handling in case the path doesn't have a period

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+        //Check to make sure key-value data has been passed with the intent
+        if (data == null) {
+            showErrorDialog(Localization.get("no.print.data"));
+        }
+
+        this.outputPath = CommCareApplication._().getTempFilePath() + ".html";
+
+        //Check if a doc location is coming in from the Intent
+        //Will return a reference of format jr://... if it has been set
+        String path = data.getString("cc:print_template_reference");
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: What if 'data' is null?

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+    /**
+     * Unique name to use for the print job name
+     */
+    private static String jobName;
+
+    /**
+     * Used to hold an instance of the WebView object being printed, so that is it not garbage
+     * collected before the print job is created
+     */
+    private WebView mWebView;
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Unused

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 10 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: That seems like an unnecessary thing to worry about here, since i is explicitly starting at 0 and counting up.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java (hunk 13 above)
Comment (amstone326) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: Yea I thought about doing that and couldn't decide if it merited the additional code and LOE, but I'll add it.

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+    }
+
+    /**
+     * A wrapper for the default print document adapter generated for a web view, to enable
+     * implementation of a custom callback when onFinish() is called
+     *
+     * Source: http://stackoverflow.com/questions/30742051/android-printmanager-get-callback
+     */
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    class PrintDocumentAdapterWrapper extends PrintDocumentAdapter {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Would be great to move this to its own file.

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 27]
```diff
@@ -0,0 +1,310 @@
...
+        jobName = inputWithoutExtension + "_" + dateString;
+    }
+
+    /**
+     * Called when TemplatePrinterTask finishes, with a result code indicating what happened. If
+     * a .html file of the filled-out template has been created and saved successfully, proceeds
+     * with printing. Otherwise, displays the appropriate error message.
+     */
+    @Override
+    public void onFinished(int result, String problemString) {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Consider renaming to templatePopulationFinished or something that makes it clear what is finishing so that you don't need to rely on the comments to provide the context.

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 28]
```diff
@@ -0,0 +1,310 @@
...
+    /**
+     * The path to the temp file location that is written to in TemplatePrinterTask, and then
+     * read back from in doHtmlPrint()
+     */
+    private String outputPath;
+
+    /**
+     * Unique name to use for the print job name
+     */
+    private static String jobName;
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Is there a reason that this is static?

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java [hunk 29]
```diff
@@ -0,0 +1,310 @@
...
+        dialogBuilder.show();
+    }
+
+    /**
+     * Shows an alert dialog about the status of the print job.
+     * Activity will quit upon exiting the dialog.
+     *
+     * @param msg String message that should be shown on the alert
+     */
+    private void showAlertDialog(String msg) {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Should be moved into PrintDocumentAdapterWrapper class, where it is used, or into a util class if others might find this useful.

File: app/src/org/commcare/android/tasks/TemplatePrinterTask.java [hunk 30]
```diff
@@ -0,0 +1,253 @@
...
+                recentString.deleteCharAt(0);
+            }
+
+            if (c == '{') {
+                if (isBetweenMustaches) {
+                    validationErrorType = VALIDATION_ERROR_MUSTACHE;
+                    throw new RuntimeException(recentString.toString());
+                } else {
+                    i++;
+                    c = input.charAt(i);
```
Comment (phillipm) [inline app/src/org/commcare/android/tasks/TemplatePrinterTask.java]: What happens if you provided this function the following string: "foo{"?
I think this line will throw IndexOutOfBoundsException

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java (hunk 24 above)
Comment (amstone326) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: There's a check for that right above.

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java (hunk 23 above)
Comment (amstone326) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: What do you think the behavior of this should be in the case that it doesn't?

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java (hunk 24 above)
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: It doesn't disrupt the control flow of the code, so it will still reach this state.

File: app/src/org/commcare/android/util/TemplatePrinterUtils.java (hunk 23 above)
Comment (phillipm) [inline app/src/org/commcare/android/util/TemplatePrinterUtils.java]: Return the empty string. Every time getExtension is called it is followed by a check that the extension returned is equal to 'html' and if it isn't you show an error dialog. Having this return the empty string will prevent runtime crashes but otherwise maintain the same behaviour

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java (hunk 26 above)
Comment (amstone326) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: I purposefully left this here for 2 reasons: 1) It really has no use outside of this activity. 2) If it were in a separate file it would need to take in the TemplatePrinterActivity as a field, since it calls showAlertDialog, which seems unnecessarily messier.

File: app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java (hunk 28 above)
Comment (amstone326) [inline app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java]: Nope..

File: app/src/org/commcare/dalvik/preferences/CommCarePreferences.java [hunk 31]
```diff
@@ -117,6 +129,63 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                }
+            } else {
+                //No file selected
+                Toast.makeText(this, Localization.get("template.not.set"), Toast.LENGTH_SHORT).show();
+            }
+        }
+
+    }
+
+    private void showAlertDialog(String title, String message) {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/preferences/CommCarePreferences.java]: Looks like this is also in TemplatePrinterActivity; might be worth putting it into a util class.

File: app/res/values-v11/themes.xml [hunk 32]
```diff
@@ -81,6 +81,8 @@
         <item name="android:paddingLeft">@dimen/title_paddingLeft</item>
 	</style>
     
-    
+    <style name="Theme.Dialog" parent="@android:style/Theme.Holo.Light.Dialog">
```
Comment (ctsims) [inline app/res/values-v11/themes.xml]: Woah. This is a change to ~all of our dialogs that removes their title bar, right? Do we want htis?

Comment (amstone326) [inline app/res/values-v11/themes.xml]: This was Richard's change, I don't actually know why it's here. Should I remove?

Comment (amstone326) [inline app/res/values-v11/themes.xml]: @ctsims Looked into this a little more, I don't think it actually has a global effect, it's only if you set an activity's style to Theme.Dialog. But I also can't really figure out what this is meant to be doing / why it's here, so would be fine with removing?

Comment (lu16j) [inline app/res/values-v11/themes.xml]: I can't remember exactly, but I think it was to improve the styling of a file selection dialog that would pop up during the process. I'm sure the flow has changed considerably since I implemented it, so if everything looks fine, it should be OK to remove.

Comment (amstone326) [inline app/res/values-v11/themes.xml]: @lu16j Thanks for the reply! I think we figured it out-- the layout for the PrintActivity is very simple, just a small progress spinner and nothing else, so it looks like you were just trying to remove the title bar from that activity, since it would look odd.

**Comment by wpride** [comment]
:fire:
